paasta_tools/kubernetes/bin/paasta_secrets_sync.py

```diff
@@ -140,34 +144,66 @@ def main() -> None:
+        ],
+    ],
+]:
+    """
+    Generate a mapping of service -> namespace -> allowlist of secrets, e.g.
+
+    {
+        "yelp-main": {
+            "paasta": {"secret1", "secret2"},
+            "paastasvc-yelp-main": {"secret1", "secret3"},
+            "paasta-flinks": None,
+        },
+        "_shared": {
+            "paasta": {"sharedsecret1"},
+            "paastasvc-yelp-main": {"sharedsecret1", "sharedsecret2"},
+            "paasta-flinks": None,
+        }
+    }
+
+    This mapping is used by sync_all_secrets / sync_secrets:
+    sync_secrets will only sync secrets into a namespace if the allowlist is None or contains that secret's name.
+    """
+    services_to_k8s_namespaces_to_allowlist: Dict[
+        str, Dict[str, Optional[Set[str]]]
+    ] = defaultdict(dict)
+
     for service in service_list:
-        # Special handling for service `_shared`, since it doesn't actually exist
-        # Copy shared secrest to all namespaces, assuming that if a secret is declared shared
-        # the team is aware that more people can see it
         if service == "_shared":
-            services_to_k8s_namespaces[service] = set(
-                INSTANCE_TYPE_TO_K8S_NAMESPACE.values()
-            )
-            paasta_namespaces = kube_client.core.list_namespace(
-                label_selector="paasta.yelp.com/managed=true"
-            )
-            for namespace in paasta_namespaces.items:
-                services_to_k8s_namespaces[service].add(namespace.metadata.name)
+            # _shared is handled specially for each service.
             continue
 
+        config_loader = PaastaServiceConfigLoader(service, soa_dir)
+        for service_instance_config in config_loader.instance_configs(
+            cluster=cluster, instance_type_class=KubernetesDeploymentConfig
+        ):
+            secrets_used, shared_secrets_used = get_secrets_used_by_instance(
+                service_instance_config
+            )
+            services_to_k8s_namespaces_to_allowlist[service].setdefault(
+                service_instance_config.get_namespace(),
+                set(),
+            ).update(secrets_used)
+
+            if "_shared" in service_list:
+                services_to_k8s_namespaces_to_allowlist["_shared"].setdefault(
+                    service_instance_config.get_namespace(),
+                    set(),
+                ).update(shared_secrets_used)
+
         for instance_type in INSTANCE_TYPES:
             if instance_type == "kubernetes":
-                config_loader = PaastaServiceConfigLoader(service, soa_dir)
-                for service_instance_config in config_loader.instance_configs(
-                    cluster=cluster, instance_type_class=KubernetesDeploymentConfig
-                ):
-                    services_to_k8s_namespaces[service].add(
-                        service_instance_config.get_namespace()
-                    )
+                continue  # handled above.
             else:
```

Comment:
random idea: thoughts on moving removing the else to reduce a layer of indentation? the `continue` above means that it'd be safe to unconditionally do the logic inside the else now

Reply:
sure